sky/packages/sky/lib/widgets/homogeneous_viewport.dart

```diff
@@ -0,0 +1,137 @@
+
+class HomogeneousViewport extends RenderObjectWrapper {
+  HomogeneousViewport({
+    Key key,
+    this.builder,
+    this.itemExtent,
+    this.itemCount,
+    this.direction: BlockDirection.vertical,
+    this.startOffset: 0.0,
+    this.token
+  }) : super(key: key);
+
+  IndexedBuilder builder;
+  double itemExtent;
+  int itemCount;
+  BlockDirection direction;
+  double startOffset;
+  Object token;
+
+  RenderBlockViewport get renderObject => super.renderObject;
+  RenderBlockViewport createNode() => new RenderBlockViewport(direction: direction, startOffset: startOffset);
+
+  bool _layoutDirty = true;
+  List<Widget> _children;
+
+  void walkChildren(WidgetTreeWalker walker) {
+    for (Widget child in _children)
+      walker(child);
+  }
+
+  void didMount() {
+    renderObject.callback = layout;
+    renderObject.totalExtentCallback = getTotalExtent;
+    renderObject.minCrossAxisDimensionCallback = getMinCrossAxisDimension;
+    renderObject.maxCrossAxisDimensionCallback = getMaxCrossAxisDimension;
+    super.didMount();
+  }
+
+  void didUnmount() {
+    renderObject.callback = null;
+    renderObject.totalExtentCallback = null;
+    renderObject.minCrossAxisDimensionCallback = null;
+    renderObject.maxCrossAxisDimensionCallback = null;
+    super.didUnmount();
+  }
+
+  bool retainStatefulNodeIfPossible(MixedViewport newNode) {
+    retainStatefulRenderObjectWrapper(newNode);
+    if (startOffset != newNode.startOffset) {
+      _layoutDirty = true;
+      startOffset = newNode.startOffset;
+    }
+    if (itemCount != newNode.itemCount) {
+      _layoutDirty = true;
+      itemCount = newNode.itemCount;
+    }
+    if (itemExtent != newNode.itemExtent) {
+      _layoutDirty = true;
+      itemExtent = newNode.itemExtent;
+    }
+    if (direction != newNode.direction) {
+      _layoutDirty = true;
+      direction = newNode.direction;
+    }
+    if (token != newNode.token) {
+      _layoutDirty = true;
+      token = newNode.token;
+    }
+    if (builder != newNode.builder) {
+      _layoutDirty = true;
+      builder = newNode.builder;
+    }
+    return true;
+  }
+
+  // This is called during the regular component build
+  void syncRenderObject(HomogeneousViewport old) {
+    super.syncRenderObject(old);
+    if (_layoutDirty) {
+      renderObject.markNeedsLayout();
+    } else {
+      assert(old != null); // if old was null, we'd be new, and therefore _layoutDirty would be true
+      syncChildren(children, old.children);
+    }
+  }
+
+  void layout(BoxConstraints constraints) {
+    LayoutCallbackBuilderHandle handle = enterLayoutCallbackBuilder();
+    try {
+      double mainAxisExtent = direction == BlockDirection.vertical ? constraints.maxHeight : constraints.maxWidth;
+      int firstIndex;
+      if (startOffset <= 0.0) {
+        firstIndex = 0;
+      } else {
+        firstIndex = startOffset ~/ itemExtent;
+      }
+      int count;
+      if (mainAxisExtent < double.INFINITY) {
```

Comment:
Negative startOffset means that it's scrolled down so the first item is lower than the top, no? If I scroll the list to half-way down the list, is startOffset positive or negative?